### PR TITLE
Refactor cache parameters

### PR DESCRIPTION
**Pull request by cecton**
Following our discussion, this PR removes the cache parameters from cli arguments and use environment variable instead. In the future when we know what and how we would like to expose these, we can make a more permanent decision. In the meantime, this will simplify the code as we don't have to carry around more things than schema, configuration and router.

## Changes
-   Remove CLI parameters for cache control
-   Remove trait WithCaching
-   Use named cached query planner instead of trait: `ApolloRouter` should have good defaults by default
-   Move NaiveIntrospection's spawn_blocking to ApolloRouter's internal
-   Move cache pre-population to ApolloRouter's internal
-   Use async_trait for RouterFactory (simplifies mocking in test)
-   Cache pre-population ran in a background task to not block the ApolloRouter's creation

    We can't spawn a task for each and every query but we can start one task for all of them so the ApolloRouter creation is not slowed down by the cache pre-population.

Related to https://github.com/apollographql/router/discussions/189

### CODE REVIEW

**Review by garypen — APPROVED** [review]
I like the fact that we are removing WithCaching. This is a nice simplification.

File: apollo-router/src/apollo_router.rs
Comment (Geal) [inline apollo-router/src/apollo_router.rs:48]: This is smart, but we might lose the benefit of warming the cache up, since it could immediately get some queries that could have been precalculated. If we block on warming up, starting the router is slightly slower, but when reloading configuration there's no delay, we create the router before restarting the server

File: apollo-router/src/apollo_router.rs
Comment (Geal) [inline apollo-router/src/apollo_router.rs:56]: (very small nitpick) we have to ignore errors, because some of the queries that were previously in the cache might not work with the new schema

**Comment by Geal** [comment]
this makes the code much simpler, nice :)

**Review by Geal — APPROVED** [review]

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs:48]: Good point!

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs:56]: Right, I will fix this, this was copy-pasted

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs]: Fixed in 2f34fcb

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs]: Fixed in 4b16837f174c7e410519e703442ee432e8b9b92b

File: apollo-router/src/apollo_router.rs
Comment (o0Ignition0o) [inline apollo-router/src/apollo_router.rs:54]: I don't understand :/

Do we here mean we are trading of update responsiveness for overall throughput?

"Populate the new router's cache before it starts handling requests. We trade off update responsiveness in favor of request processing latency."

File: apollo-router/src/lib.rs
Comment (o0Ignition0o) [inline apollo-router/src/lib.rs:564]: 🎉

File: apollo-router/src/router_factory.rs
Comment (o0Ignition0o) [inline apollo-router/src/router_factory.rs:29]: this looks a lot like Default, which is derived above. Do we have something in mind?

**Review by o0Ignition0o — APPROVED** [review]
I really like where this is going, +1 for me.

A couple of nits / questions here and there.

We might wanna make sure we still get tracing and spans (I'm never going to be at ease until we have tests that cover that, especially when we mark functions sync / async and poke around async_trait).

**Comment by o0Ignition0o** [comment]
```---- files::tests::basic_watch stdout ----
thread 'files::tests::basic_watch' panicked at 'assertion failed: futures::poll!(watch.next()).is_pending()', apollo-router/src/files.rs:79:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
``` let's keep a close eye on this, this shouldnt happen.

File: apollo-router/src/router_factory.rs
Comment (garypen) [reply apollo-router/src/router_factory.rs:29]: I think you are right. I added new() to cover the parameters. Now they have gone, I think we could remove new() and just use default().

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs:54]: This is related to this discussion: https://github.com/apollographql/router/pull/225#discussion_r760240634

What I understand: there is no point of doing this job in a background task because the server is not stopped while the new router is being created. But @Geal thinks we might lose the benefit of warming up the cache. I didn't think about this too much to be honest. Feel free to update the comment to whatever is more clear.

**Comment by cecton** [comment]
> ```
> thread 'files::tests::basic_watch' panicked at 'assertion failed: futures::poll!(watch.next()).is_pending()', apollo-router/src/files.rs:79:9
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ``` let's keep a close eye on this, this shouldnt happen.
> ```

where do you see this?

File: apollo-router/src/router_factory.rs
Comment (cecton) [reply apollo-router/src/router_factory.rs]: Fixed in f218f49

**Comment by cecton** [comment]
> We might wanna make sure we still get tracing and spans (I'm never going to be at ease until we have tests that cover that, especially when we mark functions sync / async and poke around async_trait).

I hear you and I feel the same way. Hopefully we will solve this soon.

image

Jaeger seems to be working alright. **BUT** I had a harder time to start the server with Jaeger. In the past jaeger was enabled by default but this has been removed somehow. Is this intentional?

**Comment by cecton** [comment]
> Jaeger seems to be working alright. **BUT** I had a harder time to start the server with Jaeger. In the past jaeger was enabled by default but this has been removed somehow. Is this intentional?

Created #232 to address this

**Comment by o0Ignition0o** [comment]
> In the past jaeger was enabled by default but this has been removed somehow. Is this intentional?

I think the demo default binds to starstuff, and there is no jaeger receiver, this might be the reason.

#232 does the right thing imo, good job!

**Comment by o0Ignition0o** [comment]
> > ```
> > thread 'files::tests::basic_watch' panicked at 'assertion failed: futures::poll!(watch.next()).is_pending()', apollo-router/src/files.rs:79:9
> > note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> > ``` let's keep a close eye on this, this shouldnt happen.
> > ```
> 
> where do you see this?

https://app.circleci.com/pipelines/github/apollographql/router/594/workflows/61747ae2-f942-4d91-b581-a59d5d1583cc/jobs/3120 was the failing job, I guess I need to dig further.

File: apollo-router/src/apollo_router.rs
Comment (cecton) [reply apollo-router/src/apollo_router.rs:54]: Let me know if this is still a requirement before merging. Otherwise I will merge now

Comment (o0Ignition0o) [reply apollo-router/src/apollo_router.rs:54]: I still don't understand the message, but this by no means should block merging

Comment (Geal) [reply apollo-router/src/apollo_router.rs:54]: to explain a bit more: if we first warm up the cache, then switch to the new config, the next queries will benefit from the warmed up cache. While if we switch and warm up in background, the next queries might be blocked until the cache is primed, so there'll be a perf hit

**Pull request merged**
